### PR TITLE
WIP - Badger state DB

### PR DESCRIPTION
**Pull request by jozef-slezak**
Hello, here is small pull request that implements Badger based state DB. It is related to the issue #4681.
Hopefully, you will like this idea. If not, just reuse unit tests for your Bolt state DB.

Reasons for having Badger based state DB:
1. Badger is maintained (Bolt seems to be archived github.com/hashicorp/nomad/helper/boltdd) and Nomad is not even using https://github.com/etcd-io/bbolt)
2. Badger is crash resilient  (WAL)  see https://blog.dgraph.io/post/alice/
3. Badger is pretty fast https://blog.dgraph.io/post/badger-lmdb-boltdb/

We are still facing #1367 (and able to reproduce), therefore I am experimenting on how to tweak Nomad to avoid the corrupted state.

### CODE REVIEW

**Comment by tantra35** [comment]
@jozef-slezak good job. Please clarify the following question is Badger can shrink db file unlike boltdb?

**Comment by jozef-slezak** [comment]
@tantra35 please, have a look at DB.RunValueLogGC() 
https://github.com/dgraph-io/badger#garbage-collection

**Comment by onlyjob** [comment]
Badger is improperly [vendored by HEAD](https://github.com/hashicorp/nomad/blob/dc2d4ba96121a5acb08e90edf144cdabf67c65ba/vendor/vendor.json#L436) -- @jozef-slezak, please only ever vendor [releases](https://github.com/dgraph-io/badger/releases) by tag or corresponding commit id.
Badger is semantically versioned and we should respect that. Thanks.

**Comment by jozef-slezak** [comment]
Thank you, I will improve the vendoring.

File: client/state/state_database_badger.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+	db *badger.DB
+}
+
+// NewBadgerStateDB creates or opens an existing boltdb state file or returns an
+// error.
+func NewBadgerStateDB(stateDir string) (StateDB, error) {
+	opts := badger.DefaultOptions
+	opts.Dir = stateDir
+	opts.ValueDir = stateDir
+	opts.ValueLogLoadingMode = badgeropt.FileIO //no need for (default) mmap here
```
Comment (schmichael) [inline client/state/state_database_badger.go]: I would leave the default as the state files shouldn't be too large, and letting the OS manage paging to/from disk seems optimal.

**Comment by schmichael** [comment]
Quick update after some internal discussion: sounds like the Consul team had already investigated Badger as an alternative Raft backend, but didn't find it a compelling alternative (our raft implementation basically only needs a WAL).

Making Nomad client's state backend pluggable is unlikely to be a long-term goal. The state backend is rarely the bottleneck, and the state bugs we've encountered so far have never been the fault of boltdb.

Please feel free to experiment as Badger may prove better for state storage, but it is not a high priority for us to make any performance-related tweaks to state management.

**Comment by preetapan** [comment]
Closing this, more details in the comment from @schmichael above

**Pull request closed without merging**